Review

gcomoretto opened the pull request:
This PR has been created in order to facilitate the review of the full document.

gcomoretto committed:
First Draft

gcomoretto committed:
Added Makefile and correction

gcomoretto committed:
Added lsstdraft

gcomoretto committed:
Updates

jonathansick committed:
Configure travis deployment to dmtn-106.lsst.io

jonathansick committed:
Fixup README

jonathansick committed:
Merge branch 'u/jonathansick/config'

gcomoretto committed:
Updates

gcomoretto committed:
Merge branch 'master' of https://github.com/lsst-dm/DMTN-106

gcomoretto committed:
Clarifications

gcomoretto committed:
Spelling

gcomoretto committed:
Added conditions on weekly build acceptance

gcomoretto committed:
Cleaning

womullan committed:
intro and acronyms

womullan committed:
Depenencies

womullan committed:
many updates to text

womullan committed:
tweak on change control

womullan committed:
release candidate

womullan committed:
one run through - back porting needs work

womullan committed:
fix color

womullan committed:
First draft 3rd part section

gcomoretto committed:
Rewording

gcomoretto committed:
Added extra packages to include in the tags

womullan committed:
seciton head back

gcomoretto committed:
Merge branch 'u/gcomoretto/additions' into tickets/DM-17964

gcomoretto committed:
Added status and problems appendix

gcomoretto committed:
SW Product definition update and corrections

gcomoretto committed:
Merge branch 'u/gcomoretto/appendix' into tickets/DM-17964

gcomoretto committed:
Refined SW Product and Release definition

womullan committed:
comments from john on github

womullan committed:
Merge pull request #2 from lsst-dm/tickets/DM-17964

Tickets/dm 17964

gcomoretto committed:
Introduction improvement and proposed way forward

gcomoretto committed:
Correction

womullan committed:
typos

gcomoretto committed:
Merge branch 'uwomullanwording' into tickets/DM-17964

gcomoretto committed:
Fixes

womullan committed:
fixes

gcomoretto committed:
Small update

gcomoretto committed:
Merge branch 'u/gcomoretto/improvements' into tickets/DM-17964

gcomoretto committed:
Corrections

gcomoretto committed:
Merge branch 'tickets/DM-17964'

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+This release procedure is general and valid for all software products in the LSST Data Management subsystem.
+It is an abstraction of \citeds{SQR-016}, but it is assumed that the Jenkins builds used for the Science Pipelines can be applied in the same way to other software products. 
+
+The release engineer should be in charge of the release activities documented here.  
```
Who is the release engineer? That position is not listed in the DM org tree. I'm not sure if it means “release manager”, or “some engineer who is delegated to carry out the process” (e.g. Josh).

gcomoretto replied:
Release manager.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+\begin{tabbing}
+AUTH-NUM\= \kill
+\citeds{LDM-148} \>     DM Architecture\\
+\citeds{LDM-294} \>     DM Project Management Plan   \\
```
Probably also LDM-672, albeit not yet baselined.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+This is not the case of DM, where a software product is comprised of multiple Git packages.
+The compromised solution in this case is to have a GitHub \textit{metapackage} that identify the software product.
+All Git packages composing the software product are dependencies in a GitHub \textit{metapackage} and released at the same time.
```
(I think it's clearer to refer to Git “repositories”, rather than packages, but this is incidental.)

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+These are the most common uses of an environment. Additional scenarios can be identified.
+
+In some cases, for simplicity, the same environment is used in development, tests and operations, but this does not justify the inclusion of the environment definition in the software product itself.
```
(I found this statement quite confusing at first — I thought it meant that the environment definition is included in the software product UNLESS the same environment is used in development, test, and operation. Consider rewording to make the intention clearer.)

gcomoretto replied:
What I am saying here, is that the environment definition shall not be included in the SW product.
I rephrase.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+These elements shall follow a parallel development / release process.
+
+Test data should also not be included in a software product. However including small datasets for enabling the unit tests is not a bad practice.
```
These two statements seem to directly contradict each other. :-)

Please be clear one way or the other.

gcomoretto replied:
I remove the second sentence, and I specify that Big test datasets shall not be included...

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+The identifier is usually of the form \texttt{M.n.p}.
+
+The tag in the Github repository and the software release note is sufficient to identify the release, and therefore for a developer
+to resolve the dependencies, build the binaries and execute the software.
```
Why is the release note necessary to “resolve dependences, build the binaries and execute the software”?

gcomoretto replied:
In general, the release note is the document that identifies the SW (tag) under release in a human-readable language and easy to access and share. Moreover, the release note may contain technical details and/or narrative that fully characterize the release, and that may not always be available in the tag.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+The software release is a monolithic snapshot of the software product to be used as is, or in its corresponding binary package(s), by the downstream processes or users.
+In theory, it should be possible to identify the release with a checksum calculated automatically on the delivered package.
+That checksum can be used to identify the release installed in an environment, or included in a distribution.
```
Please clarify the comment on checksums a bit.

Specifically: if I packaged the same version of the same product for release as (say) a tarball and a Docker image, then obviously the tarball and the Docker image would have different checksums. Does that mean they are actually different releases by the above definition?

gcomoretto replied:
Docker is usually not a release packaging tool, but a distribution tool. If I package the release in a tarball, that tarball has a checksum. That tarball can be included in the docker image, and it shall be possible to check the distributed tarball corresponds to the released one.

jdswinbank replied:
Ok; forget Docker. What if I compile a certain tag with clang on my Mac, and you compile it with gcc on `lsst-dev01`. We'll get different checksums — are those different releases?

gcomoretto replied:
Right. After a parallel discussion in the Arch team, I moved this sentence to the Binary Package section.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+\subsection{Binary Packages} \label{sec:swbpkg}
+
+A binary package is a package containing executable binaries for the corresponding release.
```
Presumably not just executables, but also compiled library code, and whatever supporting scripts, documentation, etc are provided but not compiled.

jdswinbank replied:
Actually, I see this is mentioned a couple of lines later, but then it's really confusing — it seems to be a contradiction of the earlier definition. Please merge line 113 with this line.

gcomoretto replied:
OK

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+A binary package is a package containing executable binaries for the corresponding release.
+It is created by building the SW provided in the release tag.
+Binary packages can be created to support multiple platforms (such as Linux, OSX, windows) if required.
```
(macOS, not OSX)

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+In some cases, these packages do not contain compiled binaries, but executable scripts or just text files.
+
+Binaries should be generated only once, and made available for their use by dependent software or for deployment.
```
Why?

To expand: I agree that there's no need to waste electricity compiling code more often than is necessary. However, assuming the build process is reproducible (and it had better be!) then there seems to be no reason not to generate binaries as often as is convenient.

gcomoretto replied:
In the context of release management, we shall build the package only once and every process downstream shall use exactly the same package. Shall, not should, I update the text.

In other contexts, like for example development, you may rebuild it as needed, but always better to reuse the same package.

The fact that the build process has to be reproducible, does not mean that we have to rebuild the binary package all the times.

jdswinbank replied:
I agree that we don't *have* to rebuild repeatedly, but it makes no sense for this document to *prohibit* multiple builds.

gcomoretto replied:
I am not saying `shall`, but `should`.

However, I specify the context (for releases).

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+Binaries should be generated only once, and made available for their use by dependent software or for deployment.
+
+The majority of DM software is using EUPS for binary packaging. Other platforms can be considered, for example \textit{conda} or \textit{pypi}.
```
eupspkg, not EUPS. (Although this is a distinction that is lost on most of us...!)

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+The majority of DM software is using EUPS for binary packaging. Other platforms can be considered, for example \textit{conda} or \textit{pypi}.
+Software products implemented in java are using jar/war binary packaging.
+\footnote{So far there are a few technologies used to handle binary packages. It is recommended to assess and optimize them, converging to use only one.
+Conda seems to have become a standard for python based projects.}
```
(For scientific Python, maybe; I don't think it has reached critical mass in the wider community.)

gcomoretto replied:
seems...

jdswinbank replied:
Well, it doesn't seem so to me. :-)

Seems like this text has been commented in current `master`. In general, please remove rather than commenting text — Git won't forget it, so you lose nothing, and it makes the source much easier to read.

gcomoretto replied:
I remove the commented text.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+\begin{itemize}
+\item to make available software releases for operations or commissioning
+\item to test (integration, validation, operation rehearsals) software releases
+\item to provide software releases to external collaborators. In this case source code can be distributed instead of binaries.
```
If a distribution is not necessarily a collection of binary packages, please fix the definition at line 124 so the document doesn't contradict itself.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+Distributions may also include non binary packages or source code, as needed.
+
+The main DM software product currently being distributed is the Science Pipeline Stack -
+this may be used in a few ways see \url{https://pipelines.lsst.io/install/index.html}
```
“may be used”, or “may be distributed”?

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+
+\subsection{Versioning and Naming} \label{sec:versioning}
+DM will  switch to Semantic Versioning \footnote{\url{https://semver.org/spec/v2.0.0.html}},  this will help improve dependency management, and make  managing patch releases more comprehensible.
```
Specifically Semver 2.0.0, or will we track new Semver releases?

gcomoretto replied:
2.0.0

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+DM software is currently versioned using two digits separated by a "." .
+For example Science Pipelines latest release:\\
+$17.0$
```
The latest release is 17.0.1.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+where \textbf{M} is the major version and \textbf{n} is the minor version starting from 0.
+
+The letter  \textbf{b} indicates that this is a branch.
+The letter \textbf{x} is a placeholder. It indicates that versions in a specific branch will have a fixed \textbf{M.n} but different patch level, where \textbf{0} means no patch.
```
(I'm confused about what the `x` is adding here, but I don't think it's a big deal)

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+All issues to be included in a release shall be added as blocking to the release issue.
+
+Note that by introducing the field \textit{Fix in Version(s)}, will permit having the same information in a much simpler way.
+Both release issue and \textit{Fix in Version(s)} fields can be used in parallel.
```
I disagree with the above. Using “Fix in Version” complicates our existing workflow and adds no value here. Certainly, using it *as well* as the current process would be a disaster!

gcomoretto replied:
"Fix in Version(s)" is a native Jira field, and it is not just about which issues block a release, but characterize an issue in terms of in which release it has been fixed.

Note that not all issues included in a release are marked as blocking for that release. Many issues may have been fixed into master before the release process starts and there will be no automatic way from Jira to know in which release they have been fixed.

Note also that a release issue will be probably needed also to trace the activity. 

Finally, developers should not be bothered about the introduction of this field, since in the vast majority of the cases it should be filled a posteriori. We should not change the workflow, at least not for this reason.

jdswinbank replied:
Git already tells you which issues have been merged into a particular release. This field adds no value given our workflow; it's only more complexity for no reason.

gcomoretto replied:
We can drop the point if this makes confusion since it is not on the critical path for improving the release process.

However, from a management point of view, this field is important.
All relevant information relevant for an issue shall be in the issue itself, without the need to go to the repository to see in which version it has been fixed.
This shall not represent any extra effort for the developer and its update shall be automatic.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+\item Installation instructions. This is usually a manually written set of instructions.
+\item Narrative section describing the content of the release, this summary should be written by a person not auto generated.
+The T/CAM is responsible for making sure this is provided.
+\item List of jira issues included in a release. This information can be extracted from Github. Completed epics will be highlighted while other issues will  listed to be comprehensive.
```
How can we extract a list of completed epics from GitHub?

One can imagine a process which tries to guess whether epics have been completed by matching tickets merged (identified on GitHub) with the contents of epics (identified on Jira), but this will be complex and error-prone.

Better just to ask people, I suggest. :-)

gcomoretto replied:
How to do it can be discussed later.. and yes, for the moment just manual collection should be fine.

jdswinbank replied:
I don't think we should publish this document unless we know it is possible to implement its recommendations!

gcomoretto replied:
Fine. However, the process to collect manually the issues included in a release is already done and is not a problem. So this is not a recommendation.

If we want to improve this aspect,  we can do it in the future.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+\begin{itemize}
+\item Installation instructions. This is usually a manually written set of instructions.
+\item Narrative section describing the content of the release, this summary should be written by a person not auto generated.
+The T/CAM is responsible for making sure this is provided.
```
(To be clear, this is the T/CAM of the product being released.)

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+Changes to the release plan  thus are  at least partially controlled the LSST change control  process as  milestones are moved or impacted by other changing milestones. Other changes not covered by the milestones such as features or content need to be proposed to the DMCCB using RFC Jira issues.
+
+Patch release requests should use an RFC Jira issue which must be approved by the DMCCB.
```
Since the development team is also impacted, the patch release should be agreed by the responsible T/CAM.

(Of course, since the DMPM is on the DM-CCB, it is always possible for the him to simply order the T/CAM to carry out a release, but I hope that will never happen.)

gcomoretto replied:
The DM-CCB process and release process defined in LDM-294 shall already take this aspect into account.

jdswinbank replied:
That's fine; so we should make sure this document does not contradict LDM-294.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+\item build downstream dependencies: CI build on SW products that depend from the newly build SW product.
+\end{itemize}
+
+At the moment, continuous integration is done on a ticket branch prior to merging changes to master. Binary packages (EUPS) and docker images for distribution are made available with the nightly and weekly builds (see next section \ref{sec:weekly}).
```
To be clear, is the above suggesting that Jenkins should _no longer_ run on ticket branches, or that it should run on _both_ ticket branches and on `master`? If the former, that sounds like a problem: developers should be able to test their changes before they merge them

gcomoretto replied:
So far I am not proposing to change Jenkins approach in that respect.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+Nightly builds do not generate tags in the Github repositories.
+When a weekly build is done, a corresponding tag in the Github repository is created.
+Though weekly builds are considered releases in Github, from a release management point of view they are not
+releases - they are reference builds and, when they meet the criteria in \ref{sec:firstrc}, release candidate starting points.
```
Are nightly & weekly builds different from simply periodically dropping a tag on the the Jenkins-on-`master` process described above?

gcomoretto replied:
This is not relevant from a release process point of view. However, the only difference that I am able to spot is that weekly builds trigger the doc build, the nightly don't.

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+\subsection{Other Related Artifacts}
+
+Together with the creation of the release candidate, other artifacts may need to be branched, and possibly have a first release candidate also:
```
I'm confused about the use of “may” here — under what circumstances would we not want to branch the documentation, environment, etc, to go with the release?

gcomoretto replied:
I will remove the "may".

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+
+\subsection{Release Candidates Validation} \label{sec:rcvalidation}
+
+The release candidate needs to be validated.
```
Who is responsible for this activity?

gcomoretto replied:
I try to add some text to clarify.
I think that T/CAM and product owner shall be responsible for this, but I am open to other suggestions.
For sure, the release manager can't validate the release, apart from ensuring that the release process has been followed.

jdswinbank replied:
Release manager can, or can't?

gcomoretto replied:
can't

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+In case problems are found during the validation, a DM issue needs to be created in Jira.
+This issue shall follow the usual development process as described in the developer guide.
+The fix will be first implemented in a ticket branch, reviewed and merged to master.
+Once the fix has been proved to work, it can be backported to the release branch.
```
Of course, it may be that the problem is discovered in a code that no longer exists in `master` (e.g. because it was deleted after the RC was tagged). In that case, I assume the fix is simply merged to the release branch.

gcomoretto replied:
This is clarified lather: "In a special case, that an issue cannot be fixed on master, the ticket branch can be opened based on the release branch."

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+\item Backport the fix from master to the release branch using the following command:
+  \begin{itemize}
+   \item $git rebase --onto bM.n.x$.
+  \end{itemize}
```
To be clear, this is simply rebasing commits; it is not backporting. Having rebased, it is possible that the rebased code will need to be edited to work correctly with the release branch.

(And being pedantic, this is backporting from `tickets/DM-XXX`, not from `master`, which may have changed since `tickets/DM-XXX` was merged).

gcomoretto replied:
I try to be more clear. I add a bullet that conflicts need to be resolved in case there are some.

jdswinbank replied:
Even resolving conflict may not be adequate — it may be, for example, that ticket DM-XYZ applies cleanly to the release branch, but simply doesn't work because it relies on ticket DM-ABC which also doesn't exist on the release.

I think suggesting specific git commands here is always going to be difficult, because there are too many possible scenarios to cover. I think it's better not to be so specific about what the developer has to type.

timj replied:
I'm happy with moving these commands to the dev guide as a "example scenarios for backporting to prior releases" section or something.

gcomoretto replied:
I am happy to remove the example from here.

gcomoretto replied:
@timj, would you like me to open a Jira issue for this update of the developer guide?

jdswinbank commented on body.tex:
```diff
@@ -0,0 +1,440 @@
+\begin{itemize}
+\item Create the backporting Jira issue DM-XYZ with subject \textit{Backport issue DM-XXX on release branch bM.n.x}
+  \begin{itemize}
+  \item DM-XXX Jira issue has to be already in status reviewed or done, and the corresponding ticket branch already merged to master.
```
DM-XXX doesn't have to have been completed in order to create the backporting issue, surely?

gcomoretto replied:
I think that if DM-XXX is in status "Reviewed", it can be backported. However, if you prefer we can state that DM-XXX need to be in status "Done" before it can be backported.

jdswinbank commented on party3.tex:
```diff
@@ -0,0 +1,35 @@
+\section{Incorporating third party code } \label{sec:party3}
```
I've struggled a bit with this section, because I'm not sure how it fits into this document.

Broadly, I think the material on third party code on which we depend (starlink_ast, for example) simply duplicates and restates established policy without adding anything beyond the idea that the DM-CCB should approve future dependencies. I think having it listed here is confusing, and I'd prefer to remove it.

The other material, about including code which builds upon our stack (rather than which our stack is built upon) seems more appropriate for LDM-672 than for this document.

Certainly, little of the below seems related to the release process, so I'm not sure why it's here.

gcomoretto replied:
I think that that topic deserves a specific document.
LDM-294 may be a good placeholder for that also, or at least mention the topic.

jdswinbank replied:
I agree with you. So are you removing it from this document?

gcomoretto replied:
I can try to move it to LDM-294, if @womullan agrees.

jdswinbank commented on status_and_problems.tex:
```diff
@@ -0,0 +1,132 @@
+\subsubsection{Distribution System} \label{sec:statusDistrib}
+
+The main tool used for distribution to operations and science community is \textbf{Docker}. 
+
```
I think this is only true if we regard deployment on the LSP as “distribution to the science community”.

I suspect the vast majority of science users who download and run code on their own systems are not using Docker.

gcomoretto replied:
Understood, I rephrase it.

jdswinbank commented on status_and_problems.tex:
```diff
@@ -0,0 +1,132 @@
+Moving 3rd party libraries to conda environment, will reduce this problem. 
+However this requires proper management of the conda environment.
+
+\textbf{Requirement}: the number of Git packages shall be kept low. DM-CCB shall approve each time a new package is introduced, since this has an impact on the maintainability of the system. 
```
Can we quantify “low”? I can't tell if this section is simply suggesting that we should avoid adding new packages in future, or if it's advocating that existing packages be merged or eliminated.

gcomoretto replied:
It is difficult to quantify. It depends on the use case. If you want to have the Alert Product software available quickly in case of fix, you cant have 120 git packages to release, because this will take 12 hours minimum, and you may want to have the fix available in 30 minutes... get this as an example. Such a problem may not exist for DRP software instead.

jdswinbank commented on status_and_problems.tex:
```diff
@@ -0,0 +1,132 @@
+\item increases the build time: each Git package needs to be build each time.
+\item increases the release time: all Git packages need to be released each time.
+\item increases the failure probability: the tooling may be affected by network glitch or similar technical issues. This may lead to the failure of the build/release process in a non deterministic way.
+\end{itemize}
```
I don't think the case has been made that any of these issues, except _maybe_ the third, are significant.

gcomoretto replied:
In the context of doing a release, all these problems are valid. At the moment.

jdswinbank replied:
I think you need to justify this. There may be some management advantages to consolidating the code into a smaller number of repositories, but why would it make a substantial difference to build time?

gcomoretto replied:
Build and unit test time may be proportional to the amount of code in the repository. However, there are other activities done each time a build is done, like for example prepare the repository. I agree this time may not be significant, it is usually few seconds per repository, but in the case of science pipelines, where now we have more than 100 repositories, this starts to be a considerable time. 
In addition, each repository needs to be packaged and pushed to eups.lsst.codes. 

So more are the repositories to consider, and more time will take the complete build process.

For sure, if we have only 10 packages, adding one will not be an issue.
But letting the number of packages grow without any control, this will be an issue.

jdswinbank commented on todo.tex:
```diff
@@ -0,0 +1,24 @@
+All the git packages included in a software product metapackage will be built and released all at the same time, as it is done now for the full Science Pipelines.
+
+A metapackage which depends on a library metapackage will resolve the library binaries from the EUPS binary repository.
+Dependencies between metapackages shall be handled differently, for example using a \textit{requirements.txt} file, than the usual EUPS requirements that are specified the the \textit{ups/gitpackagename.table}.
```
I don't understand this. Does it imply that using EUPS to `setup Calibration` will not set up the “library metapackage” on which `Calibration` depends? If so, that's not acceptable.

gcomoretto replied:
I am not expecting EUPS setup to be impacted. Here I am referring on the lsstsw rebuild, that is getting now all dependencies from GitHub. A metapackage shall be retrieved from eupspkgroot instead, if available.
This paragraph is redundant and better to remove it.

jdswinbank reviewed (commented):
Overall, I am quite happy with this. 👍 

In several places, I found the wording a bit confusing or contradictory, to the extent that I wasn't sure that I actually properly understood what was being said. I have commented on those.

In just a couple of places, I think I actually disagreed. This is a lot less disagreement than I expected! 😆 Again, I have commented on these areas.

I think almost all of the material on third-party packages is out of place here: I suspect it was added just because this document was convenient, rather than because it's actually relevant. I think it should be removed, and all of the content migrated to the Developer Guide and/or LDM-672.

There are a lot of typos and minor errors throughout the document. I have restrained myself from flagging all of them, but we should do a proofreading pass before this is actually baselined.

gcomoretto committed:
Updates due to John's comments

gcomoretto commented:
Thank you, John, for all useful comments.

I think I have answered all and changed the text in many places.

I am currently reviewing spelling and possible errors.

jdswinbank committed:
Use correct generateAcronyms.py.

jdswinbank committed:
Use older lsst-texmf.

This does not (attempt) to use glossaries, which this document doesn't
currently support.

jdswinbank committed:
Merge branch 'u/swinbank/fix-review-build' into review

gcomoretto committed:
Review status_status_and_problems, added DMTN-110

gcomoretto committed:
Review naming

gcomoretto committed:
added local to the bibliography

jdswinbank committed:
Fix author list.

gcomoretto committed:
Review first part of body.tex

gcomoretto committed:
Fixes

gcomoretto committed:
Corrected heading

gcomoretto committed:
Clarification and meta.tex

gcomoretto committed:
Further clarifications

gcomoretto committed:
Distribution clarification

gcomoretto committed:
Rephrasing and build naming added

gcomoretto committed:
SW product deinifition improvement

gcomoretto committed:
Improvements appendix A.2.3

gcomoretto committed:
Review

yalsayyad commented on body.tex:
```diff
@@ -0,0 +1,481 @@
+Some SW Products are libraries, therefore are not operationally but used as dependencies by other software products.
+\end{itemize}
+
+A software product shall correspond to a single repository.
```
Since this is a DMTN, I'd find the wording more clear if it were:

"With the exception of DM, a software product shall..." so that the DM readers know not to object when they read it. 

+ a parallel intro in the next paragraph:

gcomoretto replied:
I would like to give here a general definition.
In fact, I have removed all references to DM implementation except this one.

I would prefer to keep it generic and move the following paragraph in a footnote.

We can also change the `shall` with a `should`, in order to make less strong this statement.

What do you think?

yalsayyad commented on body.tex:
```diff
@@ -0,0 +1,481 @@
+A software product shall correspond to a single repository.
+In other words, there shall be a correspondence \textit{1:1} between software products and Git repositories.
+
+This is not the case of DM, where a software product is comprised of multiple Git repositories.
```
Parallel into above with "In DM, a software product is comprised..."

yalsayyad commented on body.tex:
```diff
@@ -0,0 +1,481 @@
+All Git repositories composing the software product are enumerated in a Git \textit{meta-package} and released at the same time.
+In this case, a Git package shall be related only to one software product, and it is not by himself a software product.
+In other words, there is a correspondence \textit{1:many} between the software product and Git repositories.
+There shall never be a \textit{many:many} correspondence.
```
Can you give an explicit example? 
afw is in lsst_apps, lsst_distrib, and lsst_sims.
The lsst_* are the software products metapackages?  afw is in many metapackages, and metapackages have many git repos

gcomoretto replied:
I rephrase the paragraph in order to be more general. This is not the right place for the proposal since it shall just be a definition section.

The only example I can give so far is lsst_distrib, that is the only `product` that we are releasing so far. 
lsst_apps is part of lsst_distrib, therefore released altogether.
lsst_sims is not released.

yalsayyad commented on body.tex:
```diff
@@ -0,0 +1,481 @@
+This is not the case of DM, where a software product is comprised of multiple Git repositories.
+The compromised solution, in this case, is to have a Git \textit{meta-package} that identifies the software product.
+All Git repositories composing the software product are enumerated in a Git \textit{meta-package} and released at the same time.
+In this case, a Git package shall be related only to one software product, and it is not by himself a software product.
```
-> "In this case, a Git package shall correspond to only one software product, and it is itself a software product."

gcomoretto replied:
This sentence is removed now.

yalsayyad commented on body.tex:
```diff
@@ -0,0 +1,481 @@
+The following elements should not be part of a software product, but follow a parallel development/release process:
+
+\begin{itemize}
+\item build tools, otherwise we will have them deployed in operations.
```
example?

gcomoretto replied:
In this moment, we are tagging as part of the lsst_distrib software product, the packages `scons` and `sconsUtils`.

In operations (Data Processing) at NCSA, we shall not deploy those packages. Scons could probably be resolved from conda and included in the environment.

Of course, those packages need to be included in the distribution to the science community.

yalsayyad commented on party3.tex:
```diff
@@ -0,0 +1,35 @@
+
+A distribution is defined in \secref{sec:distribution} an example distribution would be the docker container available as the kernel in the notebook aspect of the science platform. A collaboration may wish to get some  non LSST code included in that distribution.
+
+There are several reasons to include third part code in our distributions:
```
third-party
(Also I think third-party has a hyphen when it's used as an adjective)

yalsayyad commented on party3.tex:
```diff
@@ -0,0 +1,35 @@
+\begin{enumerate}
+\item There are packages which we use in our code e.g. starlink\_ast.  These are standard dependencies.  Currently these are put in lsst github org and probably should not be. Apart from that this is a well understood reason for third party code. \label{item:depend}
+\item There are packages which we want to use when working with the deployed software. Pandas  might be such an example - we don't need it to build or run our code but its sure handy for interaction with the data.\label{item:want}
+\item A collaborator has a package which depends on our code and is intended for some form of post processing. Another example here would be an algorithm we want to use such as ngmix which we do not necessarily need to support. This is what most people on LSST are thinking about under the topic of third party software. \label{item:colab}
```
I don't agree with "This is what most people on LSST are thinking about under the topic of third-party software." This sentence doesn't really add anything either. Can we talk about this?

gcomoretto replied:
This section is probably going to be moved in a different document.

yalsayyad commented on body.tex:
```diff
@@ -0,0 +1,481 @@
+
+The porting can be applied also in case that an issue, implemented on master after the first release candidate has been created, has to be included in the release.
+
+The DMCCB has to overview the backporting process and take corrective actions if needed. 
```
On HSC, we need a weekly call between the testers in NAOJ and the dev team in Princeton to identify and discuss the importance of specific issues for backporting

gcomoretto replied:
T/CAMs shall be involved by the DMCCB.

yalsayyad commented on body.tex:
```diff
@@ -0,0 +1,481 @@
+\item create a release candidate, for example \textit{17.0.1.rc1} using the the same approach explained above (\ref{sec:newrc})
+\item validate the release candidate
+\item fix eventual problems found in the release candidate and repeat the last 2 steps, create a new release candidate a validate it, until a valid release candidate is found
+\item create a final release as described above (\ref{sec:finalrelease}).
```
And presumably, have some additional standards like "No API Changes, etc..."

gcomoretto replied:
Right, but that should be part of the DMCCB approval of the patch.
In that case, a `regular` release should be done from master.

I am not sure this is the right place where to say this.
Maybe the 3.6 subsection in LDM-294 needs to be expanded a little bit.

gcomoretto committed:
Answers to Yusra comments

jdswinbank commented:
This work was reviewed and changes have been made.
